Lib/test/test_typing.py

```diff
@@ -2758,6 +2758,16 @@ def x(self): ...
         with self.assertRaisesRegex(TypeError, only_classes_allowed):
             issubclass(1, BadPG)
 
+        for proto in P, PG, BadP, BadPG:
+            with self.subTest(proto=proto.__name__):
+                self.assertIsSubclass(proto, Protocol)
+
+        # TODO: the behaviour of some of these is questionable!
+        # For now, just check that these don't raise any Exceptions
+        issubclass(object, Protocol)
+        issubclass(str, Protocol)
+        issubclass(C, Protocol)
```

Comment:
Some of these appear to be evaluating to `True`, which doesn't seem correct. I didn't investigate this too much, however, as this appears to be longstanding behaviour that I can reproduce on Python 3.8-3.11.

Reply:
Although, perhaps there's not much point in fixing this regression if the results of `issubclass(X, Protocol)` can't be depended upon.

Reply:
Which "some of these" is this referring to?

```
Python 3.11.0 (v3.11.0:deaf509e8f, Oct 24 2022, 14:43:23) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Protocol
>>> issubclass(object, Protocol)
False
>>> issubclass(str, Protocol)
False
>>> class C: pass
... 
>>> issubclass(C, Protocol)
False
```

Reply:
Try running this test file:

<details>

```py
import unittest

from typing import TypeVar, runtime_checkable, Protocol

class ProtocolTest(unittest.TestCase):
    def assertIsSubclass(self, cls, class_or_tuple, msg=None):
        if not issubclass(cls, class_or_tuple):
            message = '%r is not a subclass of %r' % (cls, class_or_tuple)
            if msg is not None:
                message += ' : %s' % msg
            raise self.failureException(message)

    def assertNotIsSubclass(self, cls, class_or_tuple, msg=None):
        if issubclass(cls, class_or_tuple):
            message = '%r is a subclass of %r' % (cls, class_or_tuple)
            if msg is not None:
                message += ' : %s' % msg
            raise self.failureException(message)
    
    def test_protocols_issubclass(self):
        T = TypeVar('T')

        @runtime_checkable
        class P(Protocol):
            def x(self): ...

        @runtime_checkable
        class PG(Protocol[T]):
            def x(self): ...

        class BadP(Protocol):
            def x(self): ...

        class BadPG(Protocol[T]):
            def x(self): ...

        class C:
            def x(self): ...

        self.assertIsSubclass(C, P)
        self.assertIsSubclass(C, PG)
        self.assertIsSubclass(BadP, PG)

        no_subscripted_generics = (
            "Subscripted generics cannot be used with class and instance checks"
        )

        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(C, PG[T])
        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(C, PG[C])

        only_runtime_checkable_protocols = (
            "Instance and class checks can only be used with "
            "@runtime_checkable protocols"
        )

        with self.assertRaisesRegex(TypeError, only_runtime_checkable_protocols):
            issubclass(C, BadP)
        with self.assertRaisesRegex(TypeError, only_runtime_checkable_protocols):
            issubclass(C, BadPG)

        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(P, PG[T])
        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(PG, PG[int])

        only_classes_allowed = r"issubclass\(\) arg 1 must be a class"

        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, P)
        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, PG)
        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, BadP)
        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, BadPG)

        self.assertNotIsSubclass(object, Protocol)
        self.assertNotIsSubclass(str, Protocol)
        self.assertNotIsSubclass(C, Protocol)
        for proto in P, PG, BadP, BadPG:
            with self.subTest(proto=proto.__name__):
                self.assertIsSubclass(proto, Protocol)


if __name__ == "__main__":
    unittest.main()
```

</details>

I also can't reproduce it in the REPL, which makes me think it's probably something cursed to do with a cache somewhere again.

Reply:
Indeed. I simplified it to this:

<details>
```python
import unittest

from typing import runtime_checkable, Protocol

class ProtocolTest(unittest.TestCase):
    def assertNotIsSubclass(self, cls, class_or_tuple, msg=None):
        if issubclass(cls, class_or_tuple):
            message = '%r is a subclass of %r' % (cls, class_or_tuple)
            if msg is not None:
                message += ' : %s' % msg
            raise self.failureException(message)
    
    def test_protocols_issubclass(self):
        @runtime_checkable
        class P(Protocol):
            def x(self): ...

        class C:
            def x(self): ...
        self.assertNotIsSubclass(C, Protocol)


if __name__ == "__main__":
    unittest.main()
```
</details>

On 3.11 this fails, but if you remove the unused Protocol it passes.

Reply:
Even smaller repro:

```pycon
>>> import typing
>>> @typing.runtime_checkable
... class Foo(typing.Protocol): pass  # empty protocol, so `object` is considered a subclass
...
>>> issubclass(object, typing.Protocol)
True
```

Reply:
(I'm working on a fix)